operator/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -112,390 +107,319 @@ type RedpandaReconciler struct {
 	}
 
-	// add finalizer if not exist
-	if !controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-		patch := client.MergeFrom(rp.DeepCopy())
-		controllerutil.AddFinalizer(rp, FinalizerKey)
-		if err := r.Client.Patch(ctx, rp, patch); err != nil {
-			log.Error(err, "unable to register finalizer")
-			return ctrl.Result{}, errors.WithStack(err)
-		}
-	}
-
-	// Upgrade checks. Don't reconcile if UseFlux is true or if ChartRef is set.
-	if rp.Spec.ChartRef.UseFlux != nil && *rp.Spec.ChartRef.UseFlux {
-		log.Error(nil, "useFlux: true is no longer supported. Please downgrade or unset `useFlux`")
-		return ctrl.Result{}, nil
+	// grab our existing and desired pool resources
+	// so that we can immediately calculate cluster status
+	// from and sync in any subsequent operation that
+	// early returns
+	injectedConfigVersion := ""
+	if rp.Annotations != nil && rp.Annotations[RestartClusterOnConfigChangeKey] == "true" {
+		injectedConfigVersion = rp.Status.ConfigVersion
 	}
-
-	if rp.Spec.ChartRef.ChartVersion != "" {
-		msg := "Specifying chartVersion is no longer supported. Please downgrade or unset `chartRef.chartVersion`"
-		log.Error(nil, msg)
-		r.EventRecorder.Eventf(rp, "Warning", redpandav1alpha2.EventSeverityError, msg)
-		return ctrl.Result{}, nil
+	pools, err := r.LifecycleClient.FetchExistingAndDesiredPools(ctx, rp, injectedConfigVersion)
+	if err != nil {
+		log.Error(err, "fetching pools")
+		return ctrl.Result{}, err
 	}
 
-	if err := r.reconcileResources(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
-	}
+	status := lifecycle.NewClusterStatus()
+	status.Pools = pools.PoolStatuses()
 
-	if err := r.reconcileStatus(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	if pools.AnyReady() {
+		status.Status.SetReady(statuses.ClusterReadyReasonReady)
+	} else {
+		status.Status.SetReady(statuses.ClusterReadyReasonNotReady, "No pods are ready")
 	}
 
-	if err := r.reconcileLicense(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	// Examine if the object is under deletion
+	if !rp.ObjectMeta.DeletionTimestamp.IsZero() || !isRedpandaManaged(ctx, rp) {
+		// clean up all dependant resources
+		if deleted, err := r.LifecycleClient.DeleteAll(ctx, rp); deleted || err != nil {
+			return r.syncStatusErr(ctx, err, status, rp)
+		}
+		if controllerutil.RemoveFinalizer(rp, FinalizerKey) {
+			if err := r.Client.Update(ctx, rp); err != nil {
+				log.Error(err, "updating cluster finalizer")
+				// no need to update the status at this point since the
+				// previous update failed
+				return ignoreConflict(err)
+			}
+		}
+		return ctrl.Result{}, nil
 	}
 
-	if err := r.reconcileClusterConfig(ctx, rp); err != nil {
-		return ctrl.Result{}, errors.WithStack(err)
+	// we are not deleting, so add a finalizer first before
+	// allocating any additional resources
+	if controllerutil.AddFinalizer(rp, FinalizerKey) {
+		if err := r.Client.Update(ctx, rp); err != nil {
```

Comment:
This re-populates managed fields :P

Reply:
Yep, but we return immediately after the `Update` here. The problem is that due to using `Apply` semantics in some places and `Update` elsewhere (which we need, for example to fully remove finalizers or out-of-date status conditions), we bomb out when we attempt to issue an `Update` with managed fields set (at least that was what I was seeing in test logs).